src/process/process_set.cu

```diff
@@ -182,89 +186,86 @@ namespace micm
 
       // copy data from host memory to device memory
-      cudaMemcpy(d_rate_constants, rate_constants_data, sizeof(double) * (n_grids * n_reactions), cudaMemcpyHostToDevice);
-      cudaMemcpy(d_state_variables, state_variables_data, sizeof(double) * (n_grids * n_species), cudaMemcpyHostToDevice);
-      cudaMemcpy(d_forcing, forcing_data, sizeof(double) * (n_grids * n_species), cudaMemcpyHostToDevice);
-      cudaMemcpy(d_number_of_reactants_, number_of_reactants, sizeof(size_t) * n_reactions, cudaMemcpyHostToDevice);
-      cudaMemcpy(d_reactant_ids_, reactant_ids, sizeof(size_t) * reactant_ids_size, cudaMemcpyHostToDevice);
-      cudaMemcpy(d_number_of_products_, number_of_products, sizeof(size_t) * n_reactions, cudaMemcpyHostToDevice);
-      cudaMemcpy(d_product_ids_, product_ids, sizeof(size_t) * product_ids_size, cudaMemcpyHostToDevice);
-      cudaMemcpy(d_yields_, yields, sizeof(double) * yields_size, cudaMemcpyHostToDevice);
+      cudaMemcpy(d_rate_constants, matrixParam.rate_constants_, sizeof(double) * (matrixParam.n_grids_ * matrixParam.n_reactions_), cudaMemcpyHostToDevice);
+      cudaMemcpy(d_state_variables, matrixParam.state_variables_, sizeof(double) * (matrixParam.n_grids_ * matrixParam.n_species_), cudaMemcpyHostToDevice);
+      cudaMemcpy(d_forcing, matrixParam.forcing_, sizeof(double) * (matrixParam.n_grids_ * matrixParam.n_species_), cudaMemcpyHostToDevice);
+      cudaMemcpy(d_number_of_reactants, processSet.number_of_reactants, sizeof(size_t) * matrixParam.n_reactions_, cudaMemcpyHostToDevice);
+      cudaMemcpy(d_reactant_ids, processSet.reactant_ids, sizeof(size_t) * processSet.reactant_ids_size, cudaMemcpyHostToDevice);
+      cudaMemcpy(d_number_of_products, processSet.number_of_products, sizeof(size_t) * matrixParam.n_reactions_, cudaMemcpyHostToDevice);
+      cudaMemcpy(d_product_ids, processSet.product_ids, sizeof(size_t) * processSet.product_ids_size, cudaMemcpyHostToDevice);
+      cudaMemcpy(d_yields, processSet.yields, sizeof(double) * processSet.yields_size, cudaMemcpyHostToDevice);
+      cudaMemcpy(&(device->rate_constants), &d_rate_constants, sizeof(double*),cudaMemcpyHostToDevice); 
+      cudaMemcpy(&(device->state_variables), &d_state_variables, sizeof(double*), cudaMemcpyHostToDevice); 
+      cudaMemcpy(&(device->forcing), &d_forcing, sizeof(double*), cudaMemcpyHostToDevice);   
+      cudaMemcpy(&(device->number_of_reactants), &d_number_of_reactants, sizeof(size_t*), cudaMemcpyHostToDevice); 
+      cudaMemcpy(&(device->reactant_ids), &d_reactant_ids, sizeof(size_t*), cudaMemcpyHostToDevice); 
+      cudaMemcpy(&(device->number_of_products), &d_number_of_products, sizeof(size_t*), cudaMemcpyHostToDevice); 
+      cudaMemcpy(&(device->product_ids), &d_product_ids, sizeof(size_t*), cudaMemcpyHostToDevice); 
+      cudaMemcpy(&(device->yields), &d_yields, sizeof(double*), cudaMemcpyHostToDevice); 
 
       // total thread count == number of grid cells
       int block_size = 320;
```

Comment:
I think it is reasonable to store reactant_ids, number_of_products and product_ids and all other pointers used by both kernels in one struct. I believe the structs (jacobianDevice and forcingDevice) are global and visible to all function. However within each kernel, I still need to create an instance of struct pointer and allocate device memory in each kernel function (I learned this from CUDA documentation).  Is there a way to allocate memory in one kernel function, and re-usuable in another kernel function? I can check by searching through documentation , too.

Reply:
Through some search without actual implementation and testing, it looks like I can allocate device memory once and call the kernels in sequence(AddForcingKernel(), AddJacobianKernels() ) that using those device memory before deallocating them. However, I am not sure when or in what order AddJacobianTerms and AddForcingTerms are called in the MICM package. They do not necessarily always call in sequence. Any advice on this problem?

Reply:
My bad! Actually I am referring to the integer variable `block_size = 320` and thinking if this could be defined as a global constant somewhere and shared by all the kernel drivers.

But you raised a good point: reactant_ids, number_of_products, product_ids, etc are constant for a fixed mechanism, and should be shared by different kernels without any modification of their values. I think this is something we can do for the integration test, where we could generate a struct to store all the **device** pointers here (d_reactant_ids, d_number_of_products, etc) and pass them as an input argument for each kernel driver. What do you think? @qinatan @K20shores @mattldawson

Reply:
I already pass them as a struct (Struct CUDAProcessSetParam) to a kernel driver function(eg. AddForcingTermsKernelDriver() ).  I can try out the idea of reusing device memory in the next function and see if it works.